sphinx/locale/__init__.py

```diff
@@ -188,119 +198,152 @@ def lazy_gettext(string):
             # Language couldn't be found in the specified path
             pass
-    # guarantee translators[catalog] exists
+    # guarantee translators[(namespace, catalog)] exists
     if translator is None:
-        translator = gettext.NullTranslations()
+        translator = NullTranslations()
         has_translation = False
-    translators[catalog] = translator
+    translators[(namespace, catalog)] = translator
     if hasattr(translator, 'ugettext'):
         translator.gettext = translator.ugettext
     return translator, has_translation
 
 
-def get_translator(catalog='sphinx'):
-    # type: (unicode) -> gettext.NullTranslations
-    global translators
-    translator = translators.get(catalog)
-    if translator is None:
-        translator = gettext.NullTranslations()
-    if hasattr(translator, 'ugettext'):
-        translator.gettext = translator.ugettext
-    return translator
+def init_console(locale_dir, catalog):
+    # type: (unicode, unicode) -> None
+    """Initialize locale for console."""
+    language, _ = locale.getlocale(locale.LC_MESSAGES)  # encoding is ignored
+    return init([locale_dir], language, catalog, 'console')
+
+
+def get_translator(catalog='sphinx', namespace='general'):
+    # type: (unicode) -> NullTranslations
+    return translators[(namespace, catalog)]
+
+
+def get_translation(catalog, namespace='general'):
+    # type: (unicode, unicode) -> Callable[[unicode, *Any], unicode]
+    """Get a translation function based on the *catalog*, *locale_dir*.
+
+    The extension can use this API to translate the messages on the
+    extension::
+
+        import os
+        from sphinx.locale import get_translation
+
+        _ = get_translation(__name__)
+        text = _('Hello Sphinx!')
+
+
+        def setup(app):
+            package_dir = path.abspath(path.dirname(__file__))
+            locale_dir = os.path.join(package_dir, 'locales')
+            app.add_message_catalog(__name__, locale_dir)
+
+    With this code, sphinx searches a message catalog from
+    ``${package_dir}/locales/${language}/LC_MESSAGES/${__name__}.mo``
+    The :confval:`language` is used for the searching.
+
+    .. versionadded:: 1.8
+    """
+    def lazy_gettext(message):
+        # type: (unicode) -> unicode
+        translator = get_translator(catalog, namespace)
+        return translator.gettext(message)
+
+    def gettext(message, *args):
+        # type: (unicode, *Any) -> unicode
+        if catalog not in translators:
+            # not initialized yet
+            return _TranslationProxy(lazy_gettext, message)
+        else:
+            translator = get_translator(catalog, namespace)
+            if len(args) <= 1:
+                return translator.gettext(message)
+            else:  # support pluralization
+                return translator.ngettext(message, args[0], args[1])
+
+    return gettext
+
+
+# A shortcut for sphinx-core
+_ = get_translation('sphinx')
+__ = get_translation('sphinx', 'console')
```

Comment:
These helper function is also not documented. I know they are not "public". But they are important internal APIs. So documentation is useful for maintainers.

Reply:
done